lib/io/pwm.cpp

```diff
@@ -0,0 +1,214 @@
+  if (polarity_file_ < 0) {
+    logger_.log(core::LogLevel::kFatal, "Failed to open PWM polarity file");
+    return core::Result::kFailure;
+  }
+  // Finally get the file descriptor for the enable file
+  const std::string enable_address = pwm_address + "enable";
+  enable_file_                     = open(enable_address.c_str(), O_WRONLY);
+  if (enable_file_ < 0) {
+    logger_.log(core::LogLevel::kFatal, "Failed to open PWM enable file");
+    return core::Result::kFailure;
+  }
+  logger_.log(core::LogLevel::kDebug, "Successfully initialised PWM");
+  return core::Result::kSuccess;
+}
+
+core::Result Pwm::setDutyCycleByPercentage(const core::Float duty_cycle)
+{
+  // Ensure duty cycle is between 0 and 1 inlcusive
+  if (duty_cycle > 1.0) {
+    logger_.log(core::LogLevel::kFatal,
+                "Failed to set duty cycle, percentage cannot be greater than 1.0");
+    return core::Result::kFailure;
+  }
+  if (duty_cycle < 0.0) {
+    logger_.log(core::LogLevel::kFatal,
+                "Failed to set duty cycle, percentage cannot be less than 0.0");
+    return core::Result::kFailure;
+  }
+  const auto time_active = static_cast<std::uint32_t>(duty_cycle * current_period_);
+  return setDutyCycleByTime(time_active);
+}
+
+core::Result Pwm::setDutyCycleByTime(const std::uint32_t time_active)
+{
+  if (duty_cycle_file_ < 0) {
+    logger_.log(core::LogLevel::kFatal,
+                "Failed to find PWM duty cycle file while setting duty cycle");
+    return core::Result::kFailure;
+  }
+  // Ensure active time is less or equal to period
+  if (time_active > current_period_) {
+    logger_.log(core::LogLevel::kFatal,
+                "Failed to set duty cycle, active time cannot be greater than period");
+    return core::Result::kFailure;
+  }
+  // Avoid I/O if the duty cycle is the same
+  if (time_active == current_time_active_) {
+    logger_.log(core::LogLevel::kDebug, "Duty cycle is the same as before, skipping I/O");
+    return core::Result::kSuccess;
+  }
+  current_time_active_ = time_active;
+  char write_buffer[10];
+  sprintf(write_buffer, "%d", current_time_active_);
+  const auto num_bytes_written = write(duty_cycle_file_, write_buffer, sizeof(write_buffer));
+  if (num_bytes_written != sizeof(write_buffer)) {
+    logger_.log(core::LogLevel::kFatal, "Failed to write to PWM duty cycle file");
+    return core::Result::kFailure;
+  }
+  logger_.log(core::LogLevel::kDebug, "Successfully set PWM active time to %d", time_active);
+  return core::Result::kSuccess;
+}
+
+core::Result Pwm::setPeriod(const std::uint32_t period)
+{
+  if (period_file_ < 0) {
+    logger_.log(core::LogLevel::kFatal, "Failed to find PWM period file while setting period");
+    return core::Result::kFailure;
+  }
+  // Avoid I/O if no change is required
+  if (current_period_ == period) {
+    logger_.log(core::LogLevel::kDebug, "PWM period is already set to %d", period);
+    return core::Result::kSuccess;
+  }
+  current_period_ = period;
```

Comment:
i'm not sure about the semantics here, but i believe this should only happen once the io succeeds? otherwise, if the first attempt fails, the second one will be "successful" but the io will never go through

Reply:
shoot yeah that's bad, will change